java/ql/lib/semmle/code/java/security/WeakRandomnessQuery.qll

```diff
@@ -0,0 +1,162 @@
+/** Provides classes and predicates for reasoning about weak randomness. */
+
+import java
+import semmle.code.java.frameworks.Servlets
+import semmle.code.java.security.SensitiveActions
+import semmle.code.java.dataflow.TaintTracking
+import semmle.code.java.dataflow.ExternalFlow
+import semmle.code.java.security.RandomQuery
+
+/**
+ * The `java.util.Random` class.
+ */
+class TypeRandom extends RefType {
+  TypeRandom() { this.hasQualifiedName("java.util", "Random") }
+}
+
+/**
+ * A node representing a source of weak randomness.
+ *
+ * For example, use of `java.util.Random` or `java.lang.Math.random`.
+ */
+abstract class WeakRandomnessSource extends DataFlow::Node { }
+
+/**
+ * A node representing a call to a constructor of `java.util.Random`.
+ */
+private class JavaRandomSource extends WeakRandomnessSource {
+  JavaRandomSource() {
+    this.asExpr().getType() instanceof TypeRandom and this.asExpr() instanceof ConstructorCall
+  }
+}
+
+/**
+ * A node representing a call to one of the methods of `org.apache.commons.lang.RandomStringUtils`.
+ */
+private class ApacheRandomStringUtilsMethodAccessSource extends WeakRandomnessSource {
+  ApacheRandomStringUtilsMethodAccessSource() {
+    exists(MethodAccess ma | this.asExpr() = ma |
+      ma.getMethod()
+          .hasName([
+              "random", "randomAlphabetic", "randomAlphanumeric", "randomAscii", "randomGraph",
+              "randomNumeric", "randomPrint"
+            ]) and
+      ma.getMethod()
+          .getDeclaringType()
+          .hasQualifiedName("org.apache.commons.lang", "RandomStringUtils")
+    )
+  }
+}
+
+private class ThreadLocalRandomSource extends WeakRandomnessSource {
+  ThreadLocalRandomSource() {
+    exists(MethodAccess ma | this.asExpr() = ma |
+      ma.getMethod().hasName("current") and
+      ma.getMethod()
+          .getDeclaringType()
+          .hasQualifiedName("java.util.concurrent", "ThreadLocalRandom")
+    )
+  }
+}
+
+/**
+ * The `random` method of `java.lang.Math`.
+ */
+private class MathRandomMethodAccess extends WeakRandomnessSource {
+  MathRandomMethodAccess() {
+    exists(MethodAccess ma | this.asExpr() = ma |
+      ma.getMethod().hasName("random") and
+      ma.getMethod().getDeclaringType().hasQualifiedName("java.lang", "Math")
+    )
+  }
+}
+
+/**
+ * A type which is an implementation of `java.util.Random` but considered to be safe.
+ *
+ * For example, `java.security.SecureRandom`.
+ */
+abstract private class SafeRandomImplementation extends RefType { }
```

Comment:
Likewise, note the existence of [`RandomQuery.qll`](https://github.com/github/codeql/blob/7bea18773c5a5faa2ec1c9e9849dc7f27f784a60/java/ql/lib/semmle/code/java/security/RandomQuery.qll), which could already contain definitions that you're creating here. Consider using and adding to it if it fits your needs.